File: scripts/validate-companies.js
#!/usr/bin/env node

/**
 * Validates all company files across all portals.
 * For each company slug/link, makes a HEAD/GET request and reports:
 * - ✓ ALIVE: 200-299 (valid company page)
 * - ✗ DEAD: 404/403 (company removed or no longer lists jobs)
 * - ? UNKNOWN: 5xx, timeout, network error
 *
 * Usage:
 *   node scripts/validate-companies.js [portal]  # validate all or one portal
 *
 * Portals: greenhouse, lever, ashby, workday, oracloud, dice
 */

import { readFileSync } from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import axios from 'axios';
import pLimit from 'p-limit';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT      = path.join(__dirname, '..');
const COMPANIES = path.join(ROOT, 'app', 'companies');

const TIMEOUT_MS = 5000;
const CONCURRENCY = 20;

// ─── Portal config ────────────────────────────────────────────────────────

const portals = {
    greenhouse: {
        file:    'greenhouse/greenhouse.csv',
        builder: (slug) => `https://job-boards.greenhouse.io/${slug}`,
        desc:    'Greenhouse job board',
    },
    lever: {
        file:    'lever/lever.csv',
        builder: (slug) => `https://jobs.lever.co/${slug}`,
        desc:    'Lever job board',
    },
    ashby: {
        file:    'ashbyhq/ashby.csv',
        builder: (slug) => `https://jobs.ashbyhq.com/${slug}`,
        desc:    'Ashby job board',
    },
    workday: {
        file:    'workday/wday1.json',
        builder: (entry) => entry.link,
        type:    'json',
        desc:    'Workday job board',
    },
    oracloud: {
        file:    'oracloud/oracloud.json',
        builder: (entry) => entry.jobSearchUrl,
        type:    'json',
        desc:    'Oracle Cloud jobs',
    },
    dice: {
        file:    null,
        builder: null,
        type:    'api',
        desc:    'Dice.com (API-based, no company file)',
    },
};

// ─── Loader ──────────────────────────────────────────────────────────────

const loadCompanies = (portal) => {
    if (!portals[portal].file) return [];

    const filePath = path.join(COMPANIES, portals[portal].file);
    const content  = readFileSync(filePath, 'utf8');

    if (portals[portal].type === 'json') {
        return JSON.parse(content);
    }

    // CSV: split by newline, lowercase, filter blanks and comments
    return content
        .split('\n')
        .map(line => line.toLowerCase().trim())
        .filter(line => line.length > 0 && !line.startsWith('#'));
};

// ─── Validator ────────────────────────────────────────────────────────────

const validateUrl = async (url) => {
    try {
        const response = await axios.head(url, { timeout: TIMEOUT_MS, maxRedirects: 5 });
        return { status: response.status, alive: response.status < 400 };
    } catch (err) {
        if (err.response?.status === 405) {
            // HEAD not allowed; try GET
            try {
                const response = await axios.get(url, { timeout: TIMEOUT_MS, maxRedirects: 5 });
                return { status: response.status, alive: response.status < 400 };
            } catch {
                return { status: err.response?.status ?? 0, alive: false };
            }
        }
        return { status: err.response?.status ?? 0, alive: false, error: err.message };
    }
};

// ─── Reporter ─────────────────────────────────────────────────────────────

const reportResults = (portal, results) => {
    const alive = results.filter(r => r.alive);
    const dead  = results.filter(r => !r.alive && r.status === 404);
    const other = results.filter(r => !r.alive && r.status !== 404);

    console.log(`\n${'═'.repeat(80)}`);
    console.log(`${portal.toUpperCase()} — ${portals[portal].desc}`);
    console.log(`${'═'.repeat(80)}`);
    console.log(`Total: ${results.length} | Alive: ${alive.length} | Dead (404): ${dead.length} | Other errors: ${other.length}`);

    if (dead.length > 0) {
        console.log(`\n⚠️  DEAD COMPANIES (should remove from CSV):`);
        dead.forEach(r => {
            console.log(`  ${r.status}  ${r.slug}`);
        });
    }

    if (other.length > 0) {
        console.log(`\n❓ UNKNOWN STATUS (check manually):`);
        other.slice(0, 10).forEach(r => {
            console.log(`  ${r.status || 'ERR'}  ${r.slug}  ${r.error || ''}`);
        });
        if (other.length > 10) console.log(`  ... and ${other.length - 10} more`);
    }

    if (alive.length === results.length) {
        console.log(`\n✓ All companies alive!`);
    }
};

// ─── Main ─────────────────────────────────────────────────────────────────

const validatePortal = async (portal) => {
    const companies = loadCompanies(portal);
    if (!companies.length) {
        console.log(`\n${portal.toUpperCase()}: no companies to validate`);
        return [];
    }

    console.log(`\n🔍 Validating ${portal} (${companies.length} companies)...`);

    const limit   = pLimit(CONCURRENCY);
    const results = await Promise.all(
        companies.map(company =>
            limit(async () => {
                const slug = typeof company === 'string' ? company : (company.name || company.slug);
                const url  = portals[portal].builder(company);

                if (!url) return { slug, status: 0, alive: false, error: 'No URL' };

                const result = await validateUrl(url);
                return { slug, url, ...result };
            })
        )
    );

    return results;
};

// ─── CLI ──────────────────────────────────────────────────────────────────

const args = process.argv.slice(2);
const portalArg = args[0];

(async () => {
    console.log('Job Board Company Validator');
    console.log('Checking which companies are still active...\n');

    const toValidate = portalArg ? [portalArg] : Object.keys(portals);

    for (const portal of toValidate) {
        if (!portals[portal]) {
            console.log(`❌ Unknown portal: ${portal}`);
            continue;
        }

        if (portals[portal].type === 'api') {
            console.log(`\n${portal.toUpperCase()} — uses API-based scraping (no company file)`);
            continue;
        }

        const results = await validatePortal(portal);
        if (results.length > 0) {
            reportResults(portal, results);
        }
    }

    console.log(`\n${'═'.repeat(80)}`);
    console.log('Validation complete!');
})();

File: scripts/validate_companies.py
#!/usr/bin/env python3
# Validates all company files across all portals.
# For each company slug/link, makes a HEAD/GET request and reports:
#   ALIVE  (2xx), DEAD (404/403), UNKNOWN (5xx/timeout/error)
#
#   python scripts/validate_companies.py            # all portals
#   python scripts/validate_companies.py ashby      # single portal

import json
import os
import sys
from concurrent.futures import ThreadPoolExecutor, as_completed

import requests

TIMEOUT     = 5
CONCURRENCY = 20
COMPANIES   = os.path.join('app', 'companies')

PORTALS = {
    'greenhouse': {
        'file':    'greenhouse.csv',
        'builder': lambda slug: f'https://job-boards.greenhouse.io/{slug}',
    },
    'lever': {
        'file':    'lever.csv',
        'builder': lambda slug: f'https://jobs.lever.co/{slug}',
    },
    'ashby': {
        'file':    'ashby.csv',
        'builder': lambda slug: f'https://jobs.ashbyhq.com/{slug}',
    },
    'workday': {
        'file':    'workday.json',
        'type':    'json',
        'builder': lambda entry: entry.get('link'),
    },
    'oracloud': {
        'file':    'oracloud.json',
        'type':    'json',
        'builder': lambda entry: entry.get('jobSearchUrl'),
    },
}


def load_companies(portal):
    cfg      = PORTALS[portal]
    file_path = os.path.join(COMPANIES, cfg['file'])
    with open(file_path) as f:
        content = f.read()
    if cfg.get('type') == 'json':
        return json.loads(content)
    return [l.lower().strip() for l in content.splitlines() if l.strip() and not l.startswith('#')]


def validate_url(url):
    try:
        r = requests.head(url, timeout=TIMEOUT, allow_redirects=True)
        if r.status_code == 405:
            r = requests.get(url, timeout=TIMEOUT, allow_redirects=True)
        return r.status_code, r.status_code < 400
    except requests.RequestException as e:
        return 0, False


def validate_portal(portal):
    companies = load_companies(portal)
    if not companies:
        print(f'\n{portal.upper()}: no companies to validate')
        return []

    print(f'\nValidating {portal} ({len(companies)} companies)...')
    builder = PORTALS[portal]['builder']

    def check(company):
        slug = company if isinstance(company, str) else company.get('name') or company.get('slug', '')
        url  = builder(company)
        if not url:
            return {'slug': slug, 'url': url, 'status': 0, 'alive': False}
        status, alive = validate_url(url)
        return {'slug': slug, 'url': url, 'status': status, 'alive': alive}

    results = []
    with ThreadPoolExecutor(max_workers=CONCURRENCY) as pool:
        futures = {pool.submit(check, c): c for c in companies}
        for future in as_completed(futures):
            results.append(future.result())
    return results


def report(portal, results):
    alive   = [r for r in results if r['alive']]
    dead    = [r for r in results if not r['alive'] and r['status'] == 404]
    other   = [r for r in results if not r['alive'] and r['status'] != 404]

    print(f"\n{'═' * 80}")
    print(f"{portal.upper()}")
    print(f"{'═' * 80}")
    print(f"Total: {len(results)} | Alive: {len(alive)} | Dead (404): {len(dead)} | Other errors: {len(other)}")

    if dead:
        print('\nDEAD COMPANIES (should remove from CSV):')
        for r in dead:
            print(f"  {r['status']}  {r['slug']}")

    if other:
        print('\nUNKNOWN STATUS (check manually):')
        for r in other[:10]:
            print(f"  {r['status'] or 'ERR'}  {r['slug']}")
        if len(other) > 10:
            print(f'  ... and {len(other) - 10} more')

    if len(alive) == len(results):
        print('\nAll companies alive!')


args        = sys.argv[1:]
portal_arg  = args[0] if args else None
to_validate = [portal_arg] if portal_arg else list(PORTALS.keys())

print('Job Board Company Validator')
print('Checking which companies are still active...')

for portal in to_validate:
    if portal not in PORTALS:
        print(f'Unknown portal: {portal}')
        continue
    results = validate_portal(portal)
    if results:
        report(portal, results)

print(f"\n{'═' * 80}")
print('Validation complete!')
